X86/X86MachineInstructionRaiserUtils.cpp

```diff
@@ -1068,6 +1068,19 @@ bool X86MachineInstructionRaiser::handleUnpromotedReachingDefs() {
   return true;
 }
 
+// Insert an unreachable instruction at the end of every BB if it is not
+// terminated.
+// We assume that unterminated BBs will never exit
+bool X86MachineInstructionRaiser::handleUnterminatedBlocks() {
+  LLVMContext &Ctx(getRaisedFunction()->getContext());
+  for (auto &BB : *getRaisedFunction()) {
+    if (BB.getTerminator() == nullptr) {
```

Comment:
I think insertion of any necessary `unreachable` instructions at the end of a basic block needs to be handled at the block-level. In other words, such instructions need be materialized as the basic block is populated with raised instructions. I am concerned that the proposed global loop that unconditionally inserts these would cover up potential unhandled cases.

Reply:
Currently, mctoll crashes when such a case is encountered, as the `UnifyFunctionExitNodesPass` assumes `BB.getTerminator() != nullptr` for all blocks. That was the reason why I thought putting this in a global loop was fine, as it wouldn't break any existing programs.

Alternatively, I could put this logic inside this loop:
https://github.com/microsoft/llvm-mctoll/blob/0983976177568161c22f9882623876eaa84da36b/X86/X86MachineInstructionRaiser.cpp#L5150-L5173

I will try that and submit a new patch.

Reply:
I think it is not correct to generate `unreachable` instruction in this situation.

The unhandled pattern in the test case binary has the following properties
1. the exit block of a function ends with a `call` instruction.
2. the function has a `void` return.

It appears that due to the later property, compiler optimizes the `return` instruction that would have followed `call exit_with_msg` instruction (with no real change to execution semantics). However, `llvm-mctoll` needs to materialize and insert the `return`, instead of the proposed `unhandled` instruction. Adding the `return` instruction is also in line with the function signature `void @exit_with_msg(i64 %arg1)` in the raised output and while adding `unhandled` is not.

One potential option to recognize this code pattern and insert the `return` instruction is while raising it in `raiseCallMachineInstr()` similar to what is done for tail calls in that function.

Reply:
I think inserting an unreachable instruction would be better since clang/LLVM is doing the same when compiling from C to LLVM: https://godbolt.org/z/jdazP7nEz

The generated LLVM IR for the testcase above:

```ll
define dso_local void @exit_with_msg(i8* nocapture readonly %0) local_unnamed_addr #0 !dbg !8 {
  call void @llvm.dbg.value(metadata i8* %0, metadata !16, metadata !DIExpression()), !dbg !17
  %2 = tail call i32 @puts(i8* nonnull dereferenceable(1) %0), !dbg !18
  tail call void @exit(i32 0) #4, !dbg !19
  unreachable, !dbg !19
}

declare dso_local void @exit(i32) local_unnamed_addr #1

define dso_local i32 @main() local_unnamed_addr #0 !dbg !20 {
  tail call void @exit_with_msg(i8* getelementptr inbounds ([5 x i8], [5 x i8]* @.str.1, i64 0, i64 0)), !dbg !24
  unreachable, !dbg !25
}
```

In the case of `main` inserting a return is tricky: `exit_with_msg` does not return a value, so we can't really insert a return there -- the value of `eax` is not defined there. What do you think?

Reply:
I understand. However, note that the referenced LLVM IR is generated with `-O3` which I presume includes inter-procedural analysis. That, in turn, allows the compiler to detect `exit_with_msg()` as tail call thereby leading to optimization of `return` instruction in `main()`.

1. That said, I agree that we can choose to insert `unreachable` for the kind of code pattern produced by the test in this PR.  `unreachable` instruction insertion and marking of call to exit_with_mesg()` in `main()` should be done only under certain conditions. In keeping with with my original concerns, I'd like to make sure we flag an unhandled error for other situations where unterminated basic blocks are encountered so that they can be addressed appropriately.

2. `llvm-mctoll` already generates `unreachable` instruction as terminator of `exit_with_mesg()`. It does not mark `call exit` as tail call.

I added code to address the above two aspects. Please see the attached diff [PR-152.txt](https://github.com/microsoft/llvm-mctoll/files/7833807/PR-152.txt) on top of this PR for details.

Let me know of any feedback.

Thanks for your help.

Reply:
Please use [this updated diff](https://github.com/microsoft/llvm-mctoll/files/7835198/PR-152.txt) instead of the one in my earlier reply - in case you have downloaded it. If not, I have updated the link in my earlier reply to be the same as the one in this reply, to avoid any confusion.

The additional changes check for return type of the candidate function being `void` and update comments.